src/ciphers/sm4.c

```diff
@@ -0,0 +1,489 @@
+}
+
+/*
+ * SM4 encryption.
+ */
+static void sm4_do(void *output, const void *input, const sm4_u32_t rk[32])
+{
+    sm4_u32_t Y[4];
+    sm4_u32_t X[32+4];
+    int i;
+
+    for (i = 0; i < 4; ++i)
+        X[i] = byte2u32((sm4_u8_t *)input + i * 4);
+
+    encrypt(Y, X, rk);
+    for (i = 0; i < 4; ++i)
+        u322byte((sm4_u8_t *)output + i * 4, Y[i]);
+}
+
+/*
+ * User interface
+ */
+static void sm4_encrypt(const void *input, void *output,
+                        const sm4_u32_t key[32])
+{
+    sm4_do(output, input, key);
+}
+
+/*
+ * User interface
+ */
+static void sm4_decrypt(const void *input, void *output,
+                        const sm4_u32_t key[32])
+{
+    sm4_do(output, input, key);
+}
+
+int sm4_ecb_encrypt(const unsigned char *pt, unsigned char *ct,
+                    const symmetric_key *skey)
+{
+    LTC_ARGCHK(pt   != NULL);
+    LTC_ARGCHK(ct   != NULL);
+    LTC_ARGCHK(skey != NULL);
+    sm4_encrypt(pt, ct, skey->sm4.ek);
+    return CRYPT_OK;
+}
+int sm4_ecb_decrypt(const unsigned char *ct, unsigned char *pt,
+                    const symmetric_key *skey)
+{
+    LTC_ARGCHK(pt   != NULL);
+    LTC_ARGCHK(ct   != NULL);
+    LTC_ARGCHK(skey != NULL);
+    sm4_decrypt(ct, pt, skey->sm4.dk);
+    return CRYPT_OK;
+}
+
+void sm4_done(symmetric_key *skey)
+{
+    LTC_UNUSED_PARAM(skey);
+}
+int sm4_keysize(int *keysize)
+{
+    LTC_ARGCHK(keysize != NULL);
+    if(*keysize < 16) {
+        return CRYPT_INVALID_KEYSIZE;
+    }
+    *keysize = 16;
+    return CRYPT_OK;
+}
+
+/*
+   int sm4_setup(const unsigned char *key, int keylen, int num_rounds, symmetric_key *skey);
+   int sm4_ecb_encrypt(const unsigned char *pt, unsigned char *ct, const symmetric_key *skey);
+   int sm4_ecb_decrypt(const unsigned char *ct, unsigned char *pt, const symmetric_key *skey);
+   int sm4_test(void);
+   void sm4_done(symmetric_key *skey);
+   int sm4_keysize(int *keysize);
+   */
+/*
+ * libtomcrypt interface is used
+ */
+static int sm4_self_test_ltc(void)
+{
+    int result;
+    int ret;
+    int i;
+    int keysize;
+    symmetric_key skey;
+
+    sm4_u8_t output[16];
+    sm4_u8_t plaintext[] = {
+        0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF,
+        0xFE, 0xDC, 0xBA, 0x98, 0x76, 0x54, 0x32, 0x10,
+    };
+    sm4_u8_t key[] = {
+        0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF,
+        0xFE, 0xDC, 0xBA, 0x98, 0x76, 0x54, 0x32, 0x10,
+    };
+    sm4_u8_t ciphertext[] = {
+        0x68, 0x1E, 0xDF, 0x34, 0xD2, 0x06, 0x96, 0x5E,
+        0x86, 0xB3, 0xE9, 0x4F, 0x53, 0x6E, 0x42, 0x46,
+    };
+    sm4_u8_t ciphertext_1000000t[] = {
+        0x59, 0x52, 0x98, 0xC7, 0xC6, 0xFD, 0x27, 0x1F,
+        0x04, 0x02, 0xF8, 0x04, 0xC3, 0x3D, 0x3F, 0x66,
+    };
+
+    result = 0;
+
+    sm4_printf("SM4 Self Test\n");
+
+    sm4_setup(key, sizeof(key), 32, &skey);
+
+    /*A.1 example 1*/
+    test("single encryption");
+    sm4_ecb_encrypt(plaintext, output, &skey);
+    ret = sm4_memcmp(output, ciphertext, 16);
```

Comment:
We should use `compare_testvector()` instead. Like that all the other invocations of `sm4_printf()`, `test()`, `result()` etc. can also go.